kernel/arch/dreamcast/kernel/irq.c

```diff
@@ -38,22 +38,102 @@ struct trapa_cb {
     void         *data;
 };
 
+/* Linked list of IRQ states, one is pushed onto the stack
+   every time the top-level ISR is entered. */
+struct irq_state {            // SIZE
+    bool           handled;   // 1 byte  /* mov.b only has 15-byte displacement */
+    uint8_t        code;      // 1 byte
+    uint16_t       evt;       // 2 bytes
+    volatile struct
+        irq_state *previous;  // 4 bytes
+};                            // 8 BYTES TOTAL
+
 /* Individual exception handlers */
 static struct irq_cb   irq_handlers[0x100];
-/* TRAPA exception handlers */
+/* TRAPA exception handlers. */
 static struct trapa_cb trapa_handlers[0x100];
-
-/* Global exception handler -- hook this if you want to get each and every
-   exception; you might get more than you bargained for, but it can be useful. */
+/* Global exception handler */
 static struct irq_cb   global_irq_handler;
 
 /* Default IRQ context location */
-static irq_context_t    irq_context_default;
+static irq_context_t   irq_context_default;
+/* Current IRQ state linked list pointer */
+static volatile struct irq_state *        /* Points to most recent IRQ state */
+       volatile        irq_state_current; /* Volatile pointer to volatile struct */
+
+/* Called when entiringan IRQ to push its state onto the stack as current. */
+inline static void irq_state_push(struct irq_state *current) {
+    current->previous = irq_state_current;
+    irq_state_current = current;
+}
+
+/* Called when exiting an IRQ to pop its state from the stack. */
+inline static void irq_state_pop(void) {
+    assert(irq_state_current);
+    irq_state_current = irq_state_current->previous;
+}
+
+/* Called to grab the IRQ state at a particular stack level. */
+inline static volatile struct irq_state *irq_state_n(size_t level) {
+    volatile struct irq_state *state = irq_state_current;
+    
+    for(size_t depth = 0; depth < level; ++depth) {
+        if(!state) break;
+        state = state->previous;
+    }
+
+    return state;
+}
+
+/* How deeply nested in IRQ calls are we? */
+size_t irq_int_depth(void) {
+    size_t depth = 0;
+    const irq_mask_t imask = irq_disable();
+
+    const volatile struct irq_state *state = irq_state_current;
+    
+    while(state) {
+        ++depth;
+        state = state->previous;
+    }
+
+    irq_restore(imask);
+
+    return depth;
+}
 
 /* Are we inside an interrupt? */
-static int inside_int;
-int irq_inside_int(void) {
-    return inside_int;
+bool irq_inside_int(void) {
+    return !!irq_state_current;
+}
+
+/* What's the active IRQ? */
+irq_t irq_active_int(size_t level) {
```

Comment:
Like what I mentioned elsewhere the name/comment seems incongruous with what it does. It would only be the active IRQ if level=0.

Maybe this is all because it's sort of been expanded to make way for the possibility of nesting and what's being described is what it can do now?

Reply:
Yeah, so the API is supposed to provide for the future nesting of interrupts so that there is potentially a *stack of active interrupts*. 

I think I agree, though, that maybe the comment there and on `irq_handled_int(size_t level)` and `irq_active_int(int level)` seem off, as if there is only ever a single one active, when the plan is to support multiple... updated.